spi/src/main/java/org/eclipse/dataspaceconnector/spi/asset/AssetIndex.java

```diff
@@ -15,10 +15,45 @@
 package org.eclipse.dataspaceconnector.spi.asset;
 
 import org.eclipse.dataspaceconnector.spi.types.domain.asset.Asset;
+import org.eclipse.dataspaceconnector.spi.types.domain.transfer.DataAddress;
 
 import java.util.stream.Stream;
 
+/**
+ * Query interface for {@link Asset} objects.
+ */
 public interface AssetIndex {
 
+    String FEATURE = "edc:asset:assetindex";
+
+    /**
+     * Returns all {@link Asset} objects that are selected by a certain expression
+     *
+     * @param expression An object containing a structured query to asset objects. If the expression contains no criteria,
+     *                   {@code Stream.empty()} is returned. If {@link AssetSelectorExpression#SELECT_ALL} is passed in, all
+     *                   Assets in the index are returned.
+     * @return A {@code Stream} that contains all assets that are selected by the expression. Might be empty, never null.
+     * @throws NullPointerException if the {@code AssetSelectorExpression} is null
+     */
     Stream<Asset> queryAssets(AssetSelectorExpression expression);
+
+    /**
+     * Fetches the {@link Asset} with the given ID from the metadata backend.
+     *
+     * @param assetId A String that represents the Asset ID, in most cases this will be a UUID.
+     * @return The {@link Asset} if one was found, or null otherwise.
+     * @throws NullPointerException If {@code assetId} was null or empty.
+     */
+    Asset findById(String assetId);
```

Comment:
Since we use Java 11 we should utilize its semantic capabilities. I suggest using Optional<Asset> for  obvious nullable return values.

Reply:
Hey Dominik, we discussed this at length, the codebase doesn't use Optionals in this way, and we've standardized on using nulls and annotations where appropriate (an alternative approach). That works well, and it's semantically valid in Java to do so.

Reply:
@jimmarino could worth defining a "code guideline" markdown to avoid misunderstandings?